extensions/ql-vscode/src/skeleton-query-wizard.ts

```diff
@@ -0,0 +1,247 @@
+import { CodeQLCliServer } from "./cli";
+import { OutputChannelLogger } from "./common";
+import { Credentials } from "./common/authentication";
+import { QueryLanguage } from "./common/query-language";
+import { askForLanguage, isFolderAlreadyInWorkspace } from "./helpers";
+import { getErrorMessage } from "./pure/helpers-pure";
+import { QlPackGenerator } from "./qlpack-generator";
+import { DatabaseManager } from "./local-databases";
+import * as databaseFetcher from "./databaseFetcher";
+import { ProgressCallback } from "./progress";
+
+type QueryLanguagesToDatabaseMap = Record<string, string>;
+
+export const QUERY_LANGUAGE_TO_DATABASE_REPO: QueryLanguagesToDatabaseMap = {
+  cpp: "protocolbuffers/protobuf",
+  csharp: "dotnet/efcore",
+  go: "evanw/esbuild",
+  java: "google/guava",
+  javascript: "facebook/react",
+  python: "pallets/flask",
+  ruby: "rails/rails",
+  swift: "Alamofire/Alamofire",
+};
+
+export class SkeletonQueryWizard {
+  private language: string | undefined;
+  private fileName = "example.ql";
+  private storagePath: string | undefined;
+
+  constructor(
+    private readonly cliServer: CodeQLCliServer,
+    private readonly progress: ProgressCallback,
+    private readonly credentials: Credentials | undefined,
+    private readonly extLogger: OutputChannelLogger,
+    private readonly databaseManager: DatabaseManager,
+    private readonly token: CancellationToken,
+  ) {}
+
+  private get folderName() {
+    return `codeql-custom-queries-${this.language}`;
+  }
+
+  public async execute() {
+    // show quick pick to choose language
+    this.language = await this.chooseLanguage();
+    if (!this.language) {
+      return;
+    }
+
+    this.storagePath = this.getFirstStoragePath();
+
+    const skeletonPackAlreadyExists = isFolderAlreadyInWorkspace(
+      this.folderName,
+    );
+
+    if (skeletonPackAlreadyExists) {
+      // just create a new example query file in skeleton QL pack
+      await this.createExampleFile();
+      // select existing database for language
+      await this.selectExistingDatabase();
+    } else {
+      // generate a new skeleton QL pack with query file
+      await this.createQlPack();
+      // download database based on language and select it
+      await this.downloadDatabase();
+    }
+
+    // open a query file
+    await this.openExampleFile();
+  }
+
+  private async openExampleFile() {
+    if (this.folderName === undefined || this.storagePath === undefined) {
+      throw new Error("Path to folder is undefined");
+    }
+
+    const queryFileUri = Uri.file(
+      join(this.storagePath, this.folderName, this.fileName),
+    );
+
+    try {
+      void workspace.openTextDocument(queryFileUri).then((doc) => {
+        void Window.showTextDocument(doc);
+      });
+    } catch (e: unknown) {
+      void this.extLogger.log(
+        `Could not open example query file: ${getErrorMessage(e)}`,
+      );
+    }
+  }
```

Comment:
This doesn't actually `catch` the error since we're not awaiting the result. We need to use the `.catch` method on the promise to catch this error, or `await` the result. However, I don't think we want to `await` the result since then we wouldn't be using `void` and `.then`.